### PR TITLE
Ionic audio player

### PR DESCRIPTION
**Pull request by brunokrebs**

### CODE REVIEW

File: articles/2018-05-03-title.markdown [hunk 1]
```diff
@@ -0,0 +1,1235 @@
...
+
+#### Prerequisites
+
+Since we are going to make a mobile application. We need to have the required SDKs for building the app. We are using Cordova to package our Ionic app into Native Package.
+
+* Instal the Native sdks for building application.
+
+  * For iOS applications, you need to install XCode. Check the complete requirements [here](https://cordova.apache.org/docs/en/latest/guide/platforms/ios/#installing-the-requirements). It's important to keep in mind that, you need macOS to install xCode.
+
+  * For Android applications, you need to have Android SDKs. Check the complete requirements [here](https://cordova.apache.org/docs/en/latest/guide/platforms/android/index.html#installing-the-requirements).
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: Could you, please, provide a better explanation on what readers must do? You can still point to both URLs (Android and iOS) but it would be good if you could provide a summary of what they need. Like, do they need to set the environment variables? Besides that, do they need anything else like installing SDKs and Android Studio?

File: articles/2018-05-03-title.markdown [hunk 2]
```diff
@@ -0,0 +1,1235 @@
...
+* _ClientID_ : It's unique per application. You will get it from Application's setting page.
+
+* _API Identifier_ : We need the Identifier value of API, that we have created. You will get it from API's setting page.
+
+### Scaffolding the app and Installing dependencies
+
+Let's scaffold an Ionic app using the blank template
+
+```
+$ ionic start audio-player blank
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: This command triggers a series of questions. What are the expected answers?

Comment (brunokrebs) [reply articles/2018-05-03-title.markdown]: Q1: Would you like to integrate your new app with Cordova to target native iOS and Android?

Comment (brunokrebs) [reply articles/2018-05-03-title.markdown]: Q2: Install the free Ionic Pro SDK and connect your app?

File: articles/2018-05-03-title.markdown [hunk 3]
```diff
@@ -0,0 +1,1235 @@
...
+
+```
+$ cd audio-player
+$ npm install --save @angular/animations @ngrx/store moment auth0-js @auth0/cordova @auth0/angular-jwt
+```
+
+We also need to install cordova plugins for authentication using auth0.
+
+```
+# replace {YOUR_PACKAGE_ID} with your app identifier and YOUR_AUTH0_DOMAIN with your Auth0 Domain
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: What is my app identifier?

File: articles/2018-05-03-title.markdown [hunk 4]
```diff
@@ -0,0 +1,1235 @@
...
+
+1.  To emit value, we call observer.next method with the value.
+2.  In case of an error, we fire the observer.error with error and observable ends.
+3.  If we are done with observer and there is no value to emit, we call observer.complete method.
+
+We also return a function inside Observable, which is called when Observable is unsubscribed. So can we do the clean up of our Observable in that function.
+
+We are gonna be using this observable to notify about Media Events like playing, pause, timeupdate and so on. So basically, we will listen to the media event's of Audio() inside the observable and then notify it via observer.next method. Something like following:
+
+```typescript
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: Is this some code that the reader must add to some file? If so, where are the instructions on where to add?

File: articles/2018-05-03-title.markdown [hunk 5]
```diff
@@ -0,0 +1,1235 @@
...
+      audioObj.removeEventListener('plause', handler);
+    }
+  });
+}
+```
+
+Let's add some helper functions to add and remove event listeners in a bit elegant way.
+
+```typescript
+let addEvents = function(events, obj, handler) {
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: it would be better to add some more context on where to add these code snippets...

for example, if readers were supposed to add this inside a function in a TypeScript class, I would write like this:

```typescript
// ... import statements ...

class AudioService {
  // ... constructor and other functions ...

  functionNameXYZ() {
    let addEvents = function(events, obj, handler) {
      events.forEach(event => {
        obj.addEventListener(event, handler);
      });
    };
  }
}
```

File: articles/2018-05-03-title.markdown [hunk 6]
```diff
@@ -0,0 +1,1235 @@
...
+  events.forEach(event => {
+    obj.addEventListener(event, handler);
+  });
+};
+```
+
+```typescript
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: same comment as above (more context)

File: articles/2018-05-03-title.markdown [hunk 7]
```diff
@@ -0,0 +1,1235 @@
...
+  });
+};
+```
+
+Whenever we want to play a new audio file. We want to create this observable and listen to all these media events. We will do this via playStream() method. It looks like following:
+
+```typescript
+playStream(url) {
+    return this.streamObservable(url).pipe(takeUntil(this.stop$));
+}
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: same comment as above (more context)

File: articles/2018-05-03-title.markdown [hunk 8]
```diff
@@ -0,0 +1,1235 @@
...
+```
+
+It's important to notice that we automatically unsubscribe to this Observable if **this.stop$** emits any value. We are doing this via **takeUntil** method. It literally means that the Observable is going to emit value until **this.stop$** emits no value.
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: Instead of formatting code with ** (like `this.stop$`), use back ticks, so they look like `this.stop$` (inline code). The same formatting style applies to variable names, file names, class names, functions, library names, etc, that are included in paragraphs.

File: articles/2018-05-03-title.markdown [hunk 9]
```diff
@@ -0,0 +1,1235 @@
...
+
+Let's briefly understand redux.
+
+In Redux, we manage our state in one place. In other words, we have just one object which stores/saves the state of our application. If at any point, we want to update the state, we **dispatch** and **Action** to a function, called as reducer function. This reducer function than update the state based on the action type.
+
+#### Creating Reducer
+
+By default, NgRx Action only has type property. But we need to send some information along with the type of action. So, We are going to extend the NgRx Action to suit our needs. We will name it _MediaAction_.
+
+Let's create a file _store.ts_ in _/src/providers/store/_ folder and do the following:
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: Like in here. The best format would be:

> Let's create a file called `store.ts` in the `./src/providers/store/` folder and do the following:

File: articles/2018-05-03-title.markdown [hunk 10]
```diff
@@ -0,0 +1,1235 @@
...
+```typescript
+export interface MediaAction extends Action {
+  type: string;
+  payload?: any;
+}
+```
+
+Then we need the action type for Media Events like canplay, playing and so on. Here is how we do it.
+
+```typescript
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: This is another example, with the current instructions, I have no idea at all where I have to add these. I mean, I think I have to add in the `store.ts` file, but I don't know if I'm supposed to add it before or after anything?

File: articles/2018-05-03-title.markdown [hunk 11]
```diff
@@ -122,15 +122,18 @@ a library built to integrate RxJS and Angular applications to help you manage th
...
 
-To install these libraries, you can use the following commands:
 
 ```bash
 cd audio-player
-npm install --save @angular/animations @ngrx/store moment auth0-js @types/auth0-js @auth0/cordova @auth0/angular-jwt
+npm install --save @angular/animations @ngrx/store moment auth0-js @types/auth0-js rxjs@6.2.1 rxjs-compat@6.2.1
 ```
 
+> **Note**: As of writing this article, when you create a new Ionic application. It is shipped with Angular 5.x and RxJS 5.x. But the latest `@ngrx/store` version(6.x)  requires RxJS 6.x. That's why you need to install version 6.x of RxJS. Also, Angular 5.x uses RxJS 5.x APIs. So to have backward compatibility with it, you also need to install `rxjs-compat`.
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: I would rephrase the whole paragraph as follows:

> **Note**: In the command above, you installed both `rxjs@6.2.1` and `rxjs-compat@6.2.1` because Ionic (at least, at the time of writing) ships with Angular 5 and because Angular 5 uses RxJS 5 APIs.

File: articles/2018-05-03-title.markdown [hunk 12]
```diff
@@ -122,15 +122,18 @@ a library built to integrate RxJS and Angular applications to help you manage th
 * [`moment.js`](https://momentjs.com/): a library that helps manipulating dates and times in JavaScript;
 * [`auth0-js`](https://www.npmjs.com/package/auth0-js): the official Auth0 library for JavaScript apps;
 * [`@auth0/cordova`](https://github.com/auth0/auth0-cordova): the official Auth0 library for Cordova apps;
-* [`@auth0/angular-jwt`](https://www.npmjs.com/package/@auth0/angular-jwt): the official Auth0 library to help handling JWTs on Angular apps;
+* [`rxjs`](https://github.com/ReactiveX/rxjs): a reactive programming library for JavaScript;
+* [`rxjs-compat`](https://github.com/ReactiveX/rxjs/tree/master/compat): a package to get backwards compatibility with RxJS previous to version 6;
+<!--* [`@auth0/angular-jwt`](https://www.npmjs.com/package/@auth0/angular-jwt): the official Auth0 library to help handling JWTs on Angular apps;-->
 
-To install these libraries, you can use the following commands:
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: Why did you removed this?

File: articles/2018-05-03-title.markdown [hunk 13]
```diff
@@ -153,13 +156,13 @@ Also, calling `Observable.create` will return a function that you can call when
...
 
-Now that you understand why do you need an observer, you can start by creating a service in your Ionic app:
+Now that you understand why do you need an observable, you can start by creating a service in your Ionic app:
 
 ```bash
 ionic generate provider audio
 ```
 
-This will generate a file called `audio.ts` under `./src/providers/audio/`. Replace the contents of this file with:
+This will generate a service in file called `audio.ts` under `./src/providers/audio/` and this service will be added to `NgModule` in `app.module.ts`. Replace the contents of this file with:
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: > This will generate a service in file...

to

> This will generate a service in a file...

File: articles/2018-05-03-title.markdown [hunk 14]
```diff
@@ -265,7 +268,34 @@ export class AudioProvider {
 
 ## Reading the Music Files
 
-@imsingh, add instructions here on how to set up your mock `CloudProvider`. Please, use the second-person singular narrative.
+Now that you have created Audio Service for playback features. You will need to create a service to get list of audio files.
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: Replace this whole line with:


> Now that you have created the audio service for the playback features, you will need to create a service to get a list of files. To do so, you can create a cloud service using Ionic:

File: articles/2018-05-03-title.markdown [hunk 15]
```diff
@@ -265,7 +268,34 @@ export class AudioProvider {
...
+
+You can create a cloud service using Ionic:
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: Then, remove this line (it was added in the paragraph above).

File: articles/2018-05-03-title.markdown [hunk 16]
```diff
@@ -265,7 +268,34 @@ export class AudioProvider {
...
+
+```bash
+ionic generate provider cloud
+```
+It will generate a service in file called `cloud.ts` under `./src/providers/cloud`. Replace the content of this file with:
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: Replace with:

> This command will generate a service in a file called `cloud.ts` under `./src/providers/cloud`. Now, replace the contents of this file with:

File: articles/2018-05-03-title.markdown [hunk 17]
```diff
@@ -265,7 +268,34 @@ export class AudioProvider {
...
+
+```typescript
+
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: remove this blank line

File: articles/2018-05-03-title.markdown [hunk 18]
```diff
@@ -265,7 +268,34 @@ export class AudioProvider {
...
+import { Injectable } from '@angular/core';
+import { of } from 'rxjs';
+
+@Injectable()
+export class CloudProvider {
+  files:any = [
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: Would it be able to add two files in this array?

File: articles/2018-05-03-title.markdown [hunk 19]
```diff
@@ -265,7 +268,34 @@ export class AudioProvider {
...
+      name: 'Perfect by Ed Sheeran'
+    }
+  ];
+  getFiles() {
+   return of(this.files);
+  }
+}
+```
+
+The `getFiles` method above basically mocks a `http` request by returning an `Observable` with a static `files` object.
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: > The `getFiles` method above basically mocks an HTTP request by returning an `Observable` with a static `files` object.

HTTP is not a command nor a filename, rather a proper noun (the name of a protocol). That's why you don't need to add back ticks.

File: articles/2018-05-03-title.markdown [hunk 20]
```diff
@@ -1160,8 +1190,7 @@ The final steps, before being able to use your app for the first time, starts wi
 import { BrowserModule } from '@angular/platform-browser';
 import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
 import { ErrorHandler, NgModule } from '@angular/core';
-import { HttpClientModule } from '@angular/common/http';
-import { JwtModule, JWT_OPTIONS } from '@auth0/angular-jwt';
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: Are you removing Auth0 from this prototype? Or are you still refactoring?

Comment (brunokrebs) [reply articles/2018-05-03-title.markdown]: Oh, never mind. Just realised you don't need it.

File: articles/2018-05-03-title.markdown (hunk 11 above)
Comment (imsingh) [reply articles/2018-05-03-title.markdown]: It is much cleaner. But it doesn't tell why we need RxJS 6.x. Thing is, we you install @ngrx/store.  It will install the latest version of it(6.x). That version of NgRx is dependent on RxJS 6. Can you try to add that as well in it?

File: articles/2018-05-03-title.markdown [hunk 21]
```diff
@@ -0,0 +1,1265 @@
...
+  }
+}
+```
+
+Within each `case` statement in the code above, you are generating a new `state` to your app. It's important to note that, as NgRx works with `immutable` objects, you need to create a new state object instead of updating the existing one. In this case, you are using `Object.assign` to create the new `state` object based on the current one.
+
+Now, to register your reducer in your Ionic app, open the `app.module.ts` file and update it as follows:
+
+```typescript
+// ... other import statements ...
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: import {StoreModule} from '@ngrx/store';
import {mediaStateReducer} from '../providers/store/store';

File: articles/2018-05-03-title.markdown [hunk 22]
```diff
@@ -0,0 +1,1265 @@
...
+To develop a secure app, you are going to rely on Auth0 to handle the authentication of your users. As such, you can <a href="https://auth0.com/signup" data-amp-replace="CLIENT_ID" data-amp-addparams="anonId=CLIENT_ID(cid-scope-cookie-fallback-name)">sign up for a free Auth0 account here</a>. Then, you will need to set up an Auth0 Application to represent your mobile app.
+
+### Installing Dependencies
+
+To secure your Ionic app with Auth0, you will have to install some Cordova plugins:
+
+```bash
+# replace {YOUR_PACKAGE_ID} with your app identifier and
+# replace YOUR_AUTH0_DOMAIN with your Auth0 Domain
+ionic cordova plugin add cordova-plugin-customurlscheme --variable URL_SCHEME={YOUR_PACKAGE_ID} --variable ANDROID_SCHEME={YOUR_PACKAGE_ID} --variable ANDROID_HOST={YOUR_AUTH0_DOMAIN} --variable ANDROID_PATHPREFIX=/cordova/{YOUR_PACKAGE_ID}/callback
```
Comment (brunokrebs) [inline articles/2018-05-03-title.markdown]: break command
